src/backend/abuseDetection.ts

```diff
@@ -0,0 +1,140 @@
+  const resultKeys = flatMap(fingerprintKeys, key => timeRange.map(() => key)); // e.g. [ 'source_ip', 'source_ip', 'source_ip', 'user_agent', 'user_agent', 'user_agent', ...
+  const tallyAbuseScore = (obj: AbuseScore, key: keyof AbuseScore, num: number) => ({
+    ...obj,
+    [key]: fp[key] ? obj[key] + num : ABUSE_SCORE_MISSING, // if there was no input for this fingerprint key (e.g. no X-Forwarded-For was provided to begin with), mark the return value as missing; otherwise, add to previous value
+  });
+  const readPromise = dynamoDb.getValues(keysToGet).then(res =>
+    res.reduce(
+      (memo, next, i) => tallyAbuseScore(memo, resultKeys[i], next), // correlate each result with its corresponding fingerprint key (from resultKeys) and increment the relevant field in the score object
+      { ...emptyScore }, // start with an empty score
+    ),
+  );
+  const writePromise = readPromise.then(
+    () => Promise.all(keysToIncrement.map(dynamoDb.incrementKey)), // after we're done reading, ask DynamoDB to increment the relevant keys
+  );
+  return { readPromise, writePromise };
+}
+
+// Creates a specialized DynamoDB API wrapper for reading/writing abuse detection related data
+export function createDynamoDbClient(
+  tableName: string,
+  ttlSeconds = TIME_RANGE_HOURS * 60 * 60, // by default, expire each key 24 h after its last write
+) {
+  var ddb = new AWS.DynamoDB({ apiVersion: '2012-08-10' }); // note: for local development, you may need to: AWS.config.update({ region: 'eu-west-1' });
+  return {
+    // Increments the integer value at given key.
+    // If the key doesn't exist, it's created automatically as having value 0, then incremented normally.
+    // Returns the integer value (after being incremented).
+    incrementKey(key: string): Promise<number> {
+      return ddb
+        .updateItem({
+          TableName: tableName,
+          Key: { ADKey: { S: key } },
+          UpdateExpression: 'SET ADVal = if_not_exists(ADVal, :init) + :inc, ADTtl = :ttl',
+          ExpressionAttributeValues: {
+            ':inc': { N: '1' },
+            ':init': { N: '0' },
+            ':ttl': { N: Math.round((Date.now() + 1000 * ttlSeconds) / 1000) + '' },
+          },
+          ReturnValues: 'UPDATED_NEW',
+        })
+        .promise()
+        .then(res => res?.Attributes?.ADVal)
+        .then(unwrapNumber);
+    },
+
+    // Returns the integer values at given keys.
+    // If some keys don't contain values, they're treated as 0's.
+    getValues(keys: string[]): Promise<number[]> {
+      return ddb
+        .batchGetItem({
+          RequestItems: {
+            [tableName]: {
+              AttributesToGet: ['ADKey', 'ADVal'],
+              Keys: keys.map(key => ({ ADKey: { S: key } })),
+            },
+          },
+        })
+        .promise()
+        .then(res =>
+          (res?.Responses?.[tableName] || []).reduce(
+            (memo, next) => ({ ...memo, [next.ADKey.S || '']: unwrapNumber(next.ADVal) }),
```

Comment:
Interesting choice of naming the reducer parameters to `memo` and `next`. I always used `acc` and `value`, but these somehow look more appealing 🤔

Reply:
Not sure where I picked that up, just happened to get etched into my brain so now I write it out automatically, otherwise no chance to remember the order of args. :stuck_out_tongue: